qiskit/providers/v2/job.py

```diff
@@ -0,0 +1,140 @@
+
+    def __init__(self, backend: Backend, job_id: str, **kwargs) -> None:
+        """Initializes the asynchronous job.
+
+        Args:
+            backend: the backend used to run the job.
+            job_id: a unique id in the context of the backend used to run
+                the job.
+            kwargs: Any key value metadata to associate with this job.
+        """
+        self._job_id = job_id
+        self._backend = backend
+        self.metadata = kwargs
+
+    def job_id(self) -> str:
+        """Return a unique id identifying the job."""
+        return self._job_id
+
+    def backend(self) -> Backend:
+        """Return the backend where this job was executed."""
+        return self._backend
+
+    def done(self) -> bool:
+        """Return whether the job has successfully run."""
+        return self.status() == JobStatus.DONE
+
+    def running(self) -> bool:
+        """Return whether the job is actively running."""
+        return self.status() == JobStatus.RUNNING
+
+    def cancelled(self) -> bool:
+        """Return whether the job has been cancelled."""
+        return self.status() == JobStatus.CANCELLED
+
+    def in_final_state(self) -> bool:
+        """Return whether the job is in a final job state."""
+        return self.status() in JOB_FINAL_STATES
+
+    def wait_for_final_state(
+            self,
+            timeout: Optional[float] = None,
+            wait: float = 5,
+            callback: Optional[Callable] = None
+    ) -> None:
+        """Poll the job status until it progresses to a final state such as ``DONE`` or ``ERROR``.
+
+        Args:
+            timeout: Seconds to wait for the job. If ``None``, wait indefinitely.
+            wait: Seconds between queries.
+            callback: Callback function invoked after each query.
+                The following positional arguments are provided to the callback function:
+
+                * job_id: Job ID
+                * job_status: Status of the job from the last query
+                * job: This BaseJob instance
+
+                Note: different subclass might provide different arguments to
+                the callback function.
+
+        Raises:
+            JobTimeoutError: If the job does not reach a final state before the
+                specified timeout.
+        """
+        if not self._async:
+            return
+        start_time = time.time()
+        status = self.status()
+        while status not in JOB_FINAL_STATES:
+            elapsed_time = time.time() - start_time
+            if timeout is not None and elapsed_time >= timeout:
+                raise JobTimeoutError(
+                    'Timeout while waiting for job {}.'.format(self.job_id()))
+            if callback:
+                callback(self.job_id(), status, self)
+            time.sleep(wait)
+            status = self.status()
+        return
+
+    @abstractmethod
+    def submit(self):
+        """Submit the job to the backend for execution."""
```

Comment:
It is used by basic aer: https://github.com/Qiskit/qiskit-terra/blob/master/qiskit/providers/basicaer/qasm_simulator.py#L393 I did rip it out in the v2 migration (https://github.com/Qiskit/qiskit-terra/pull/5086/commits/9ede763d57632d593549facdac3e02d8cb75dc3b is the revert of that) because I made basic aer hardcoded as synchronous.

Reply:
It looks like the code was ported to aer too: https://github.com/Qiskit/qiskit-aer/blob/8114479b0938f26f2f69f7e091368c3ec8e8cba1/qiskit/providers/aer/backends/aerbackend.py#L111 https://github.com/Qiskit/qiskit-aer/blob/9492a71633223c451a0bc606e41d898b2c2779c4/qiskit/providers/aer/backends/pulse_simulator.py#L134